### PR TITLE
Add the possibility to deny spawning actors for dead players

### PR DESCRIPTION
**Pull request by abcdefg30**
Some maps/mods spawn player controllable actors using `SpawnActorOnDeath`. This leads to some weird results when a player is defeated (e.g. surrenders), as the player is dead, but can still control units. It makes sense to still spawn non-controllable units like husks, imho, so I made this a setting.

### CODE REVIEW

File: OpenRA.Mods.Common/Traits/SpawnActorOnDeath.cs [hunk 1]
```diff
@@ -50,6 +50,9 @@ public class SpawnActorOnDeathInfo : ITraitInfo
 			"lead to unexpected behaviour.")]
 		public readonly CVec Offset = CVec.Zero;
 
+		[Desc("Defines if an actor is spawned even after the player was defeated (e.g. after surrendering).")]
+		public readonly bool SpawnAfterGameEnd = true;
```
Comment (Phrohdoh) [inline OpenRA.Mods.Common/Traits/SpawnActorOnDeath.cs]: A player being defeated isn't necessarily game end.

Comment (abcdefg30) [inline OpenRA.Mods.Common/Traits/SpawnActorOnDeath.cs]: Well, the game ended for this one player. :P
I just didn't want to go with `SpawnAfterDeath` (death of the player), as that is kind of ambiguous. Now that I think of it again, what about `SpawnAfterDefeat`?

Comment (Phrohdoh) [inline OpenRA.Mods.Common/Traits/SpawnActorOnDeath.cs]: `SpawnAfterDefeat` 👍

I would reword the beginning of the desc to be `Should this actor spawn after the player...`.

**Comment by abcdefg30** [comment]
Updated.

**Comment by GraionDilach** [comment]
This is clearly a bug. The fix should be to shift the ownership of these actors to the world's owner immediately or skip spawning.

Noncontrollable player-owned units is mehy IMO as well (I believe husks should be ownerless in the currently shipped mods because it looks stupid when they blow up along with the player and we have the required traits and setups to change that, AS proves).

In any case, dead players should never have controllable actors.

**Comment by rob-v** [comment]
This isn't solution to issue described by Graion, though can be used as workaround to stop spawning actors of dead players. The bug could be fixed in another PR.
In case this new property SpawnAfterDefeat is useful anyway, then 👍 .

**Comment by pchote** [comment]
> Noncontrollable player-owned units is mehy IMO as well (I believe husks should be ownerless in the currently shipped mods because it looks stupid when they blow up along with the player and we have the required traits and setups to change that, AS proves).

I suggest filing a specific issue for this for discussion.  Just switching owner isn't good enough because that would change the player color, but that could be solved by implementing `IEffectiveOwner` on the husk trait.

**Comment by GraionDilach** [comment]
I can't imagine any usecase where this setting would be preferred and wouldn't considered as a bug.

The husk issue I mentioned is something which affects balance so I'm quite sure the playerbase would shoot that down. For the record, I am spawning all husks as neutral via
```
SpawnActorOnDeath:
		OwnerType: InternalName
		InternalOwner: Creeps
```

**Comment by pchote** [comment]
There are two limitations there, but easily fixable:
* Spawning an actor that is supposed to be player controllable.  This can be fixed by adding an `OwnerType.InternalNameIfVictimDefeated` and falling back to `InternalOwner` if the player is defeated.
* The player color will change.  This can be solved for `Husk` specifically by using `IEffectiveOwner` to spoof the `ParentActorInit`'s owner.  This would also fix any balance issues.

**Comment by reaperrr** [comment]
Needs rebase.

**Comment by abcdefg30** [comment]
Updated. (The defaults/yaml values might need some adjustments, but I want some comments on the changes first.)

**Comment by reaperrr** [comment]
Needs another rebase, sorry.

**Comment by abcdefg30** [comment]
Rebased.

File: OpenRA.Mods.Common/Traits/Husk.cs [hunk 2]
```diff
@@ -43,10 +43,13 @@ IEnumerable<object> IActorPreviewInitInfo.ActorPreviewInits(ActorInfo ai, ActorP
 		bool IOccupySpaceInfo.SharesCell { get { return false; } }
 	}
 
-	public class Husk : IPositionable, IFacing, ISync, INotifyCreated, INotifyAddedToWorld, INotifyRemovedFromWorld, IDeathActorInitModifier
+	public class Husk : IPositionable, IFacing, ISync, INotifyCreated, INotifyAddedToWorld, INotifyRemovedFromWorld,
+		IDeathActorInitModifier, IEffectiveOwner
 	{
-		readonly HuskInfo info;
 		readonly Actor self;
+		readonly bool isDisguised;
```
Comment (reaperrr) [inline OpenRA.Mods.Common/Traits/Husk.cs]: I'd prefer if this was made less ambigous, so nobody confuses this with disguise logic for actors.
Something like `effectiveOwnerDisguised`, even if it's a bit long.

**Comment by abcdefg30** [comment]
Updated.

File: OpenRA.Mods.Common/Traits/Husk.cs [hunk 3]
```diff
@@ -118,20 +124,23 @@ public void SetPosition(Actor self, WPos pos)
...
 			self.World.RemoveFromMaps(self, this);
 		}
 
-		public void ModifyDeathActorInit(Actor self, TypeDictionary init)
+		void IDeathActorInitModifier.ModifyDeathActorInit(Actor self, TypeDictionary init)
 		{
 			init.Add(new FacingInit(Facing));
 		}
+
+		bool IEffectiveOwner.Disguised { get { return effectiveOwnerDisguised; } }
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/Husk.cs]: Why don't you hardcode `true` here? Returning `self.Owner` from `IEffectiveOwner.Owner` shouldn't be a problem?

File: OpenRA.Mods.Common/Traits/SpawnActorOnDeath.cs [hunk 4]
```diff
@@ -29,11 +28,13 @@ public class SpawnActorOnDeathInfo : ConditionalTraitInfo
...
 		[Desc("Owner of the spawned actor. Allowed keywords:" +
-			"'Victim', 'Killer' and 'InternalName'.")]
+			"'Victim', 'Killer' and 'InternalName'. " +
+			"Falls back to 'InternalName' if 'Victim' is used " +
+			"and the victim is defeated (see 'SpawnAfterDefeat').")]
 		public readonly OwnerType OwnerType = OwnerType.Victim;
 
 		[Desc("Map player to use when 'InternalName' is defined on 'OwnerType'.")]
-		public readonly string InternalOwner = null;
+		public readonly string InternalOwner = "Creeps";
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/SpawnActorOnDeath.cs]: Why Creeps?  IMO Neutral would be a more intuitive default.

File: OpenRA.Mods.Common/Traits/SpawnActorOnDeath.cs [hunk 5]
```diff
@@ -50,6 +51,9 @@ public class SpawnActorOnDeathInfo : ConditionalTraitInfo
 			"lead to unexpected behaviour.")]
 		public readonly CVec Offset = CVec.Zero;
 
+		[Desc("Should an actor spawn after the player has been defeated (e.g. after surrendering)?")]
+		public readonly bool SpawnAfterDefeat = false;
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/SpawnActorOnDeath.cs]: Tanks not spawning husks when a player surrenders is a bug.  This should default to true!

File: OpenRA.Mods.Common/Traits/SpawnActorOnDeath.cs [hunk 6]
```diff
@@ -90,6 +94,10 @@ void INotifyRemovedFromWorld.RemovedFromWorld(Actor self)
 			if (attackingPlayer == null)
 				return;
 
+			var defeated = self.Owner.WinState != WinState.Undefined;
+			if (defeated && Info.SpawnAfterDefeat)
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/SpawnActorOnDeath.cs]: The check here is inverted.  We want to suppress the husks when `SpawnAfterDefeat` is false.

**Review by pchote — COMMENTED** [review]
This looks straightforward enough.  A couple of comments:

**Comment by pchote** [comment]
When a player surrenders we currently kill all husks that were already on the map.  Fixing that would go well with these improvements, is but probably out of scope of this PR.

**Comment by pchote** [comment]
When a player surrenders we kill all husks that were already on the map.  The easiest way to fix that is to always pass `EffectiveOwnerInit`, and change the husk defaults to always spawn as Neutral (but pretending to be the original owner).

File: OpenRA.Mods.Common/Traits/SpawnActorOnDeath.cs (hunk 4 above)
Comment (GraionDilach) [reply OpenRA.Mods.Common/Traits/SpawnActorOnDeath.cs]: RA2 spawned to Creeps.

Comment (pchote) [reply OpenRA.Mods.Common/Traits/SpawnActorOnDeath.cs]: In OpenRA creeps are considered hostile and will target (and more importantly for husks etc be targeted *by*) all players.

**Comment by pchote** [comment]
@abcdefg30: Any progress here, or should someone else take this over?  IMO this is too nice of a polish fix to close and forget about.

File: OpenRA.Mods.Common/Traits/SpawnActorOnDeath.cs (hunk 4 above)
Comment (pchote) [reply OpenRA.Mods.Common/Traits/SpawnActorOnDeath.cs]: By the principle of least surprise our defaults should be as boring as possible, and modders should have to opt in to any additional/surprising behaviour.  Neutral fits that bill much better than Creeps, and justifying creeps by saying that Husks turn off the creep specific behavior reinforces that point.

File: OpenRA.Mods.Common/Traits/Husk.cs (hunk 3 above)
Comment (abcdefg30) [reply OpenRA.Mods.Common/Traits/Husk.cs]: Why should I hardcode it to `true`? `IEffectiveOwner.Owner` already returns `self.Owner` btw.

**Comment by abcdefg30** [comment]
Updated. Sorry for the delay.

File: OpenRA.Mods.Common/Traits/Husk.cs (hunk 3 above)
Comment (pchote) [reply OpenRA.Mods.Common/Traits/Husk.cs]: Exactly because `IEffectiveOwner.Owner` returns `self.Owner`.  That method returns the correct result whether the source actor was disguised or not, so there is no need to store extra state to tell the game whether to query `self.Owner` from the actor, or `self.Owner` from `IEffective.Owner`.

File: OpenRA.Mods.Common/Traits/SpawnActorOnDeath.cs [hunk 7]
```diff
@@ -90,6 +97,10 @@ void INotifyRemovedFromWorld.RemovedFromWorld(Actor self)
 			if (attackingPlayer == null)
 				return;
 
+			var defeated = self.Owner.WinState != WinState.Undefined;
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/SpawnActorOnDeath.cs]: This should be checking `== WinState.Defeated`.  I know this is pedantic because the game stops when somebody wins, but the name should still match the implementation.

File: OpenRA.Mods.Common/Traits/SpawnActorOnDeath.cs [hunk 8]
```diff
@@ -29,11 +28,16 @@ public class SpawnActorOnDeathInfo : ConditionalTraitInfo
...
+			"Falls back to 'InternalName' if 'Victim' is used " +
+			"and the victim is defeated (see 'SpawnAfterDefeat').")]
 		public readonly OwnerType OwnerType = OwnerType.Victim;
 
 		[Desc("Map player to use when 'InternalName' is defined on 'OwnerType'.")]
-		public readonly string InternalOwner = null;
+		public readonly string InternalOwner = "Neutral";
+
+		[Desc("Changes the effective (displayed) owner of the spawned actor to the old owner (victim).")]
+		public readonly bool DisplayOldOwner = false;
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/SpawnActorOnDeath.cs]: IMO `EffectiveOwnerFromOwner` would be a more precise and clear name for this.

**Review by pchote — DISMISSED** [review]
Just a couple of nits, otherwise LGTM:

**Comment by abcdefg30** [comment]
Updated.

File: OpenRA.Mods.Common/Traits/SpawnActorOnDeath.cs [hunk 9]
```diff
@@ -98,8 +109,24 @@ void INotifyRemovedFromWorld.RemovedFromWorld(Actor self)
...
+				td.Add(new EffectiveOwnerInit(self.EffectiveOwner.Owner));
+			else if (Info.EffectiveOwnerFromOwner)
+				td.Add(new EffectiveOwnerInit(self.Owner));
+
 			if (Info.OwnerType == OwnerType.Victim)
-				td.Add(new OwnerInit(self.Owner));
+			{
+				// Fall back to InternalOwner if the Victim was defeated,
+				// but only if InternalOwner is defined
+				if (!defeated || string.IsNullOrEmpty(Info.InternalOwner))
```
Comment (GraionDilach) [inline OpenRA.Mods.Common/Traits/SpawnActorOnDeath.cs]: This sounds like able to recreating the bug the ticket is about if I just empty out InternalOwner with SpawnAfterDefeat: true.

File: OpenRA.Mods.Common/Traits/Husk.cs [hunk 10]
```diff
@@ -128,10 +132,14 @@ void INotifyRemovedFromWorld.RemovedFromWorld(Actor self)
...
 		}
 
-		public void ModifyDeathActorInit(Actor self, TypeDictionary init)
+		void IDeathActorInitModifier.ModifyDeathActorInit(Actor self, TypeDictionary init)
 		{
 			init.Add(new FacingInit(Facing));
 		}
+
+		// We return self.Owner if there's no effective owner
+		bool IEffectiveOwner.Disguised { get { return true; } }
```
Comment (penev92) [inline OpenRA.Mods.Common/Traits/Husk.cs]: I'm not sure I like the implications of "husks are always disguised".

Comment (abcdefg30) [reply OpenRA.Mods.Common/Traits/Husk.cs]: See https://github.com/OpenRA/OpenRA/pull/13276#discussion_r146107838 (and IRC).

File: OpenRA.Mods.Common/Traits/SpawnActorOnDeath.cs (hunk 9 above)
Comment (abcdefg30) [reply OpenRA.Mods.Common/Traits/SpawnActorOnDeath.cs]: It is intentional then (if you put those values), so not really an issue? (Someone might want to do it after all.)

Comment (GraionDilach) [reply OpenRA.Mods.Common/Traits/SpawnActorOnDeath.cs]: Meh, alright - https://github.com/OpenRA/OpenRA/pull/9117#discussion_r37540853 logic can be applied to this case as well.

**Review by reaperrr — APPROVED** [review]

**Comment by reaperrr** [comment]
[changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/03eba0858d1d742aa441a504c8f60c541194aea0...10722c822c08ca0b86b5e565870f84b7d4ba2f22)
